pkg/auth/authmap_cache.go

```diff
@@ -66,6 +73,32 @@ func (r *authMapCache) Delete(key authKey) error {
 	return nil
 }
 
+func (r *authMapCache) DeleteIf(predicate func(key authKey, info authInfo) bool) error {
+	r.cacheEntriesMutex.Lock()
+	defer r.cacheEntriesMutex.Unlock()
+
+	var entriesToDelete []authKey
+	for k, v := range r.cacheEntries {
+		if predicate(k, v) {
+			entriesToDelete = append(entriesToDelete, k)
+		}
+	}
+
+	for _, key := range entriesToDelete {
```

Comment:
Why wouldn't we just do a single loop that checks the predicate and deletes?

Reply:
@joamaki thought it isn't safe to delete entries of the map which gets iterated over the same time.

thanks for the hint that it is!

https://go.dev/ref/spec#For_statements

> The iteration order over maps is not specified and is not guaranteed to be the same from one iteration to the next. If a map entry that has not yet been reached is removed during iteration, the corresponding iteration value will not be produced. If a map entry is created during iteration, that entry may be produced during the iteration or may be skipped. The choice may vary for each entry created and from one iteration to the next. If the map is nil, the number of iterations is 0.